``` {r setup, echo=FALSE, message=FALSE, include=FALSE, error=FALSE}
library(GenomicRanges, warn.conflicts=F)
library(magrittr)
library(parallel)
library(plyr)
library(BSgenome.Dmelanogaster.UCSC.dm3)


setwd("/data/analysis_code/")
options(knitr.figure_dir = "half_life_extended_table")
source("shared_code/knitr_common.r")
source("shared_code/granges_common.r")
source("shared_code/metapeak_common.r")
source("shared_code/sample_common.r")
```

# Paused Pol II half life extended table

**Author:** [Wanqing Shao](mailto:was@stowers.org)

**Generated:** `r format(Sys.time(), "%a %b %d %Y, %I:%M %p")`

### column name explanation

**fb_t_id**: FlyBase transcript ID.

**fb_g_id**: FlyBase gene ID.

**gene**: Common name.

**half_life**: Paused Pol II half-life in minute. 

**rank** and **quantile**: All the promoters were ranked based on their paused Pol II half-life and seperated into 5 quantiles.

**pro_cap**: TSSs from flybase protein coding genes (fb-r5.47) were re-annotated to match a nearby PRO-cap (GSM1032759, Lis lab) peak summit if within 150 bp distance.

**dmso_polii**, **dmso_tbp**, **dmso_tfiib**, **dmso_tfiif** and **dmso_nelfe**: Pol II, TBP, TFIIB, TFIIF and NELF-E ChIP-nexus signal within a 201 bp window around the TSS. Unit: Reads per million (log2).

**DPE**, **DRE**, **Inr**, **MTE**, **PB**, **TATA**, **Motif1**, **Motif6**, **Motif7** and **TCT**: Core promoter elements. Please refer to supplementary table 2 for their consensus sequence and relative position to the TSS. 

```{r half_life_table}

tss <- get(load("rdata/dme_mrna_unique_tss.RData"))
half_life_df <- get(load("./rdata/half_life_df.RData"))

half_life_tss <- tss[tss$fb_t_id %in% half_life_df$fb_t_id]

samples <- c("dmso_polii", "dmso_tbp","dmso_tfiib", "dmso_tfiif", "dmso_nelfe")

sig_info <- cache("signal_list.rds", function(){
  lapply(samples, function(x){
  sig <- nexus_regionSums(resize(half_life_tss, 201, "center"), load_bigwig(x))
  sig_df <- data.frame(fb_t_id = half_life_tss$fb_t_id, sig = log(sig, base=2))
  colnames(sig_df)[2] <- x
  sig_df
})
})

sig_info_df <- reshape::merge_recurse(sig_info)
### generate procap sig table
procap_pos <- import("/data/public_data/lis_procap/GSM1032759_PROcap.pl.bedgraph.gz")
procap_neg <- import("/data/public_data/lis_procap/GSM1032759_PROcap.mn.bedgraph.gz")

tss_pos <- half_life_tss[strand(half_life_tss) == "+"]
tss_neg <- half_life_tss[strand(half_life_tss) == "-"]

tss_pos$procap_sig <- coverage(procap_pos, weight=abs(procap_pos$score)) %>% regionSums(resize(tss_pos, 301, "center"), .)
tss_neg$procap_sig <- coverage(procap_neg, weight=abs(procap_neg$score)) %>% regionSums(resize(tss_neg, 301, "center"), .)

procap_sig <- data.frame(fb_t_id = c(tss_pos$fb_t_id, tss_neg$fb_t_id), pro_cap = c(tss_pos$procap_sig, tss_neg$procap_sig))
procap_sig$pro_cap <- ifelse(procap_sig$pro_cap >= 20, "T", "F")
### generate promoter element table

find_motif <- function(motif_name, fb_t_id,  mismatch=0) {
    
    motif_info <- subset(promoter_table, name == motif_name)
    motif <- DNAString(motif_info$motif)
    up_dis <- motif_info$window_start
    down_dis <- motif_info$window_end
    
    gene_tss <- tss[tss$fb_t_id %in% fb_t_id]
    
    if(up_dis >= 0 & down_dis >=0){
      tss_r <- resize(gene_tss, down_dis, "start") %>%
               resize(., down_dis - up_dis, "end")
    }
    if(up_dis < 0 & down_dis >=0){
      tss_r <- resize(gene_tss, down_dis, "start") %>%
               resize(., abs(up_dis)+down_dis, "end")
    }
    if(up_dis < 0 & down_dis <0){
      tss_r <- resize(gene_tss, abs(up_dis), "end") %>%
               resize(., abs(up_dis)-abs(down_dis), "start")
    }
    
    promoter_seq <- getSeq(Dmelanogaster, tss_r)
    names(promoter_seq) <- tss_r$fb_t_id
    
    count_df <- vcountPattern(motif, promoter_seq, fixed = FALSE, min.mismatch = 0, max.mismatch = mismatch) %>%
                data.frame(fb_t_id = fb_t_id, count =.)
    
    count_df$count <- ifelse(count_df$count >0, "T", "F")
    colnames(count_df)[2] <- motif_name
    count_df
}

promoter_table <- read.table("/data/analysis_code/promoter_elements.txt", header=T)
motifs <- promoter_table$name

motif_list <- cache("motif_list.rds", function(){
  lapply(as.character(motifs), function(x){
  motif <- find_motif(motif_name=x, half_life_tss$fb_t_id)
  motif
  })
})

motif_df <- reshape::merge_recurse(motif_list)

all_info_table <- merge(half_life_df[, c(1, 2, 3, 4, 6, 7)], sig_info_df)  %>% merge(., procap_sig) %>% merge(., motif_df)
all_info_table <- all_info_table[order(all_info_table$rank), ]
all_info_table$half_life <- round(all_info_table$half_life, digits = 2)
all_info_table$half_life <- ifelse(all_info_table$half_life>0 & all_info_table$half_life<= 60, all_info_table$half_life, "> 60")
write.table(all_info_table, file="/data/analysis_code/half_life_extended_table.txt")  
rownames(all_info_table) <- NULL
pander(all_info_table, "half life extended table")
```
```{r}
sessionInfo()
```