<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title></title>

<script src="html_support/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="html_support/bootstrap-3.3.5/css/spacelab.min.css" rel="stylesheet" />
<script src="html_support/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="html_support/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="html_support/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="html_support/navigation-1.1/tabsets.js"></script>
<link href="html_support/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="html_support/highlightjs-1.1/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="html_support/custom.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="fluid-row" id="header">




</div>

<div id="TOC">
<ul>
<li><a href="#tss-re-annotation">TSS re-annotation</a><ul>
<li><a href="#overview">Overview</a><ul>
<li><a href="#generate-flybase_r5.57-mrna-transcripts-granges-object">Generate flybase_r5.57 mRNA transcripts granges object</a></li>
<li><a href="#re-align-tss-to-pro-cap-summits">Re-align TSS to PRO-cap summits</a></li>
<li><a href="#filter-out-overlapping-tss">Filter out overlapping TSS</a></li>
</ul></li>
</ul></li>
</ul>
</div>

<div id="tss-re-annotation" class="section level1">
<h1>TSS re-annotation</h1>
<p><strong>Author:</strong> <a href="mailto:was@stowers.org">Wanqing Shao</a></p>
<p><strong>Generated:</strong> Fri Apr 07 2017, 05:34 PM</p>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>Re-annotate Dme TSS using PRO-cap data from Lis (Kwak et al, 2013)</p>
<p>TSSs from flybase protein coding genes (fb-r5.47) were re-annotated to match a nearby PRO-cap (GSM1032759) peak summit if within 150 bp distance. Original TSS annotations from flybase were preserved if no PRO-cap peak was detected. If a TSS was found within 300 bp of another TSS, both TSSs were removed from the promoter set. This set of re-annotated TSSs (n= 14,229) was used for subsequent analyses.</p>
<div id="generate-flybase_r5.57-mrna-transcripts-granges-object" class="section level3">
<h3>Generate flybase_r5.57 mRNA transcripts granges object</h3>
<pre class="r"><code>fb_5.57_mrna &lt;- read.table(&quot;/data/public_data/flybase/flybase_r5.57_mrna.txt&quot;, header=F, sep = &quot;\t&quot;)

fb_5.57_mrna &lt;- fb_5.57_mrna[, c(1, 3, 4,5,7, 9)]
colnames(fb_5.57_mrna) &lt;- c(&quot;chr&quot;, &quot;type&quot;, &quot;start&quot;, &quot;end&quot;,  &quot;strand&quot;, &quot;info&quot;)
fb_5.57_mrna &lt;- subset(fb_5.57_mrna, strand %in% c(&quot;+&quot;, &quot;-&quot;))

tx.gr &lt;- with(fb_5.57_mrna, GRanges(ranges   = IRanges(start=start, end=end), 
                                 strand   = as.character(strand), 
                                 seqnames = paste0(&quot;chr&quot;, chr),
                                 fb_t_id  = gsub(&quot;.*ID=&quot;, &quot;&quot;, info) %&gt;% gsub(&quot;;.*&quot;, &quot;&quot;, .),
                                 fb_g_id  = gsub(&quot;.*Parent=&quot;, &quot;&quot;, info) %&gt;% gsub(&quot;;.*&quot;, &quot;&quot;, .),
                                 gene  = gsub(&quot;.*Name=&quot;, &quot;&quot;, info) %&gt;% gsub(&quot;-R.*&quot;, &quot;&quot;, .)))

valid_chr &lt;- c(&quot;chr2L&quot;, &quot;chr2R&quot;, &quot;chr3L&quot;, &quot;chr3R&quot;, &quot;chr4&quot;, &quot;chrX&quot; )
seqlevels(tx.gr, force=T) &lt;- valid_chr
save(tx.gr, file=&quot;/data/analysis_code/rdata/dme_mrna.gr.RData&quot;)</code></pre>
</div>
<div id="re-align-tss-to-pro-cap-summits" class="section level3">
<h3>Re-align TSS to PRO-cap summits</h3>
<pre class="r"><code>procap_pos &lt;- import(&quot;/data/public_data/lis_procap/GSM1032759_PROcap.pl.bedgraph.gz&quot;)
procap_neg &lt;- import(&quot;/data/public_data/lis_procap/GSM1032759_PROcap.mn.bedgraph.gz&quot;)


tss_realignment &lt;- function(strand, procap){
  tx &lt;- tx.gr[strand(tx.gr) == strand]
  tx.r &lt;- resize(tx, 1, &quot;start&quot;) %&gt;% resize(., 301, &quot;center&quot;)
  tx.r$new_start &lt;- coverage(procap, weight=abs(procap$score)) %&gt;% regionWhichMaxs(tx.r,.) 
  tx.r$procap_sig &lt;- coverage(procap, weight=abs(procap$score)) %&gt;% regionSums(tx.r, .)
  start(tx.r) &lt;- ifelse(tx.r$procap_sig &gt;= 20, tx.r$new_start, start(tx.r))
  end(tx.r) &lt;- start(tx.r)
  tx.r
}

tss_pos &lt;- tss_realignment(&quot;+&quot;, procap_pos)
tss_neg &lt;- tss_realignment(&quot;-&quot;, procap_neg)
tss &lt;- c(tss_pos, tss_neg)</code></pre>
</div>
<div id="filter-out-overlapping-tss" class="section level3">
<h3>Filter out overlapping TSS</h3>
<pre class="r"><code>tss_temp &lt;- tss[order(tss$procap_sig, decreasing=T)] %&gt;% 
            .[!duplicated(paste(seqnames(.), start(.)))]

strand(tss_temp) &lt;- &quot;*&quot;
tss_temp &lt;- tss_temp[order(tss_temp)]

all_dis &lt;- c(start(tss_temp), Inf) - c(0, start(tss_temp))

tss_temp$dis_before &lt;- all_dis[1:length(tss_temp)]
tss_temp$dis_after &lt;- all_dis[2:length(all_dis)]

tss_temp &lt;-subset(tss_temp, dis_before &gt; 300 &amp; dis_after &gt; 300) 
tss_u &lt;- tss[tss$fb_t_id %in% tss_temp$fb_t_id]

save(tss_u, file=&quot;/data/analysis_code/rdata/dme_mrna_unique_tss.RData&quot;)</code></pre>
<pre><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.2 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] digest_0.6.12        pander_0.6.0         rtracklayer_1.32.2  
## [4] magrittr_1.5         GenomicRanges_1.24.3 GenomeInfoDb_1.8.7  
## [7] IRanges_2.6.1        S4Vectors_0.10.3     BiocGenerics_0.18.0 
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.9                knitr_1.15.1              
##  [3] XVector_0.12.1             GenomicAlignments_1.8.4   
##  [5] zlibbioc_1.18.0            BiocParallel_1.6.6        
##  [7] stringr_1.2.0              tools_3.3.2               
##  [9] SummarizedExperiment_1.2.3 Biobase_2.32.0            
## [11] htmltools_0.3.5            yaml_2.1.14               
## [13] rprojroot_1.2              bitops_1.0-6              
## [15] RCurl_1.95-4.8             evaluate_0.10             
## [17] rmarkdown_1.3              stringi_1.1.2             
## [19] backports_1.0.5            Biostrings_2.40.2         
## [21] Rsamtools_1.24.0           XML_3.98-1.5</code></pre>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>


</body>
</html>
